### ISSUE TITLE
osh pwd and ls disagree on current dir

### ISSUE THREAD

**Issue by dezren39** [oils-for-unix/oils#2058]
when a tool such as zed opens an inline terminal,

i've encountered an issue where somehow, `ls` and `cat` see my dir as '/home/user/code', but `pwd` and `cd` see my current dir as `/home/user`.

when the new shell opens, pwd shows /home/user
```
osh-0.22.0$ pwd
/home/user
```
when i ls $PWD, it agrees and shows /home/user files
```
osh-0.22.0$ ls $PWD
 39-slip     data	        hyprland.todo	<...>
```
but plain ls shows /home/user/code files (this is the dir that zed is attempting to drop me into)
```
osh-0.22.0$ ls
a		 doc		  generate_gpg_key.sh <...>
```
cat is on board with what ls sees,
```
osh-0.22.0$ cat generate_gpg_key.sh
#!/usr/bin/env bash
<...>
```
but cd agrees with pwd, it says i'm in /home/user and so i can't relative jump to /home/user/code/infra/dns
```
osh-0.22.0$ cd infra/dns
  cd infra/dns
     ^~~~~
[ interactive ]:15: cd '/home/user/infra/dns': No such file or directory
```
if i cd into code, or anywhere really, now all is aligned.
```
osh-0.22.0$ cd code; ls
a		 doc		  generate_gpg_key.sh <...>
osh-0.22.0$ pwd
/home/user/code
```
```
osh-0.22.0$ cd infra/dns
osh-0.22.0$
```

i'm unclear how this difference occurs

**Comment by andychu** [comment]
Hmmm ... thanks for the report

There is some legacy shell logic around `$PWD pwd $HOME` and so forth that is very complex,  which could be related to this

However I thought that OSH was matching other shells very closely

----

There are some related test cases here, and the only failing one doesn't seem related?

https://www.oilshell.org/release/0.23.0/test/spec.wwz/osh-py/builtin-cd.html

---

Do you have any symlinks or anything odd about the config?  There are corner cases in shell dealing with `cd` to a symlink and so forth

**Comment by andychu** [comment]
It would be interesting if we can get a repro  with a tool other than zed?

That might help narrow it down

---

Hm actually I found this our code


```
     # TODO on $HOME: bash sets it if it's a login shell and not in POSIX mode!
     # if (login_shell == 1 && posixly_correct == 0)
     #   set_home_var ();

```

I am not sure if this is exactly related, but it seems suspect

But this is $HOME and not $PWD

Sometimes they are related but maybe not here

**Comment by dezren39** [comment]
so far i've only had it happen with zed, but i will maybe try another editor later and see if it happens there too. nothing specifically odd about my setup other than i am using nixos. but my repo was cloned into ~ and i don't use any indirection there.

**Comment by andychu** [comment]
Oh I think I  just fixed this !  On May 12th


- https://github.com/oils-for-unix/oils/commit/5fce0cee993d1bf10307c2f261a1e49e728dd9da

There is a very long "bug post mortem" here

https://oilshell.zulipchat.com/#narrow/channel/503208-bug-post-mortem/topic/Perl.20cd.20build.20failure.20.2F.20POSIX.20Rule.20about.20cached.20PWD/with/517731503 (login required)

Long story short is that most shells have a rule like this when inheriting `$PWD` from the environment -- something like it is mandated by POSIX:


dash:

```
if (p)
    if (*p != '/' || stat(p, &st1) || stat(".", &st2) ||
        st1.st_dev != st2.st_dev || st1.st_ino != st2.st_ino)
        p = 0;
```

bash:


```
   /* Follow posix rules for importing PWD */
   if (temp_var && imported_p (temp_var) &&
       (temp_string = value_cell (temp_var)) &&
       temp_string[0] == '/' &&
       same_file (temp_string, ".", (struct stat *)NULL, (struct stat *)NULL))
     {
```


So basically they need to use PWD because they **respect symlinks**

But if `$PWD` does not point at the **same inode** as `getcwd()`, then it must be **invalidated** -- i.e. ignored and recomputed

But you can't recompute it unconditionally!

**Comment by andychu** [comment]
Oh I also is ksh93 the default shell on illumos?  If so, I didn't know that

I didn't know development was active

We have many spec tests that other shells have used:

https://oils.pub/release/0.29.0/test/spec.wwz/osh-py/index.html

I learned of this Brush shell the other day, aiming for POSIX and bash compatibility - https://news.ycombinator.com/item?id=43908368

I will probably run our spec tests on it ... and I may include other shells.  Not sure if it makes sense to include ksh93 ... I don't think I have ever run ksh93 on Linux (?)  

I thought that GNU `libc` wasn't enough to run ksh93 (?), though I could be wrong

**Comment by sjohannes** [comment]
Yeah, it does look like that commit should fix this bug. Thanks!

---

illumos uses [a fork](https://github.com/illumos/ast) of the unmaintained [AT&T ksh93](https://github.com/att/ast) as their `/bin/sh` but not necessarily the default login shell (the distro I tried, OpenIndiana, uses Bash as the default login shell). I'm not sure how much maintenance illumos wants to commit to their ksh93 fork, though.

I wasn't able to build AT&T ksh93 on Linux (or the illumos fork), but Debian did manage to do it at some point. [Here are their patches](https://salsa.debian.org/debian/ksh/-/tree/debian/93u+20120801/debian/patches), if you want to try.

There's also an active ksh93 fork called [ksh93u+m](https://github.com/ksh93/ksh). That's what Debian and Fedora use for their "ksh" package now.

From the pdksh lineage, other than mksh, there is also OpenBSD ksh which has a cross-platform port called [oksh](https://github.com/ibara/oksh) and a Linux-only port called [loksh](https://github.com/dimkr/loksh).

**Comment by andychu** [comment]
Ah I was unaware of `ksh93u+m` !  https://github.com/ksh93/ksh?tab=readme-ov-file

That is kind of a weird name ...

> Between 2017 and 2020 there was an ultimately unsuccessful [attempt](https://github.com/att/ast/tree/2020.0.1) to breathe new life into the KornShell by extensively refactoring the last unstable AST beta version (93v-). While that ksh2020 effort is now abandoned and still has many critical bugs, it also had a lot of bugs fixed.

----

Our spec tests are run against `mksh`, which is the shell on Android, derived from `pdksh`

I didn't realize that Debian and Fedora have `ksh` packages ....

---

My picture of things is

- bash aims to be ksh-compatible, but bash also has many weird behaviors and bugs.  It is underspecified IMO
- mksh also aims to be ksh-compatible, but I think it's not as strong as an interactive shell.  It doesn't seem to receive much maintenance

and

- OSH is tested against bash and mksh, so it is quite compatible with both.  
  - Although I guess mksh arrays are different than bash arrays?  I am not sure how ksh93 arrays compare.

If there is active development going on in ksh, then I guess it's because people on Illumos already have many ksh scripts?  And maybe they are running Linux-derived shell scripts on Illumos, and need more features?

As I understand it, that is basically how busybox ash development goes -- there is "pressure" to run bash scripts.

---

I think OSH should work on Illumos, but I haven't tried it.  It works on OS X and FreeBSD.

The requirements are only a C++ compiler, and a POSIX shell to build it (no make or cmake)


Nine Reasons to Use OSH - https://oils.pub/osh.html

One issue I know about is that the upcoming bash 5.4 is implementing a ksh feature `${ echo hi; }`, so we may implement that at some point

---

In the next year, we will publish a compatibility report focused on bash and busybox ash, because those are very popular shells

We may also include `ksh93u+m`, if that is interesting.  Since there is a Debian package, it should be trivial to run our spec tests with it

**Comment by andychu** [comment]
### Shell compatibility reports - ksh passes ~1460 tests

@sjohannes I'm not sure where I would send this, but a few other shell authors found this comparison interesting!

https://pages.oils.pub/spec-compat/2025-06-26/renamed-tmp/spec/compat/TOP.html

https://pages.oils.pub/spec-compat/2025-06-26/renamed-tmp/spec/compat/PASSING.html

e.g. showing that bash implemented a lot of  `mksh` and `ksh` , but also has many more features

I'm not sure if bash compatibility is a goal of modern `ksh` development or not?

**Comment by sjohannes** [comment]
Thanks, perhaps the ksh93u+m people might find it useful (they have a [discussion board](https://github.com/ksh93/ksh/discussions) on GitHub).

**Comment by andychu** [comment]
Fixed quite awhile ago - this was a tricky bug - https://oils.pub/blog/2025/09/releases.html

Thank you for the report!
